src/enclave/tls_endpoint.h

```diff
@@ -220,13 +220,14 @@ namespace enclave
     struct SendRecvMsg
     {
       std::vector<uint8_t> data;
-      std::shared_ptr<Endpoint> self;
+      std::weak_ptr<Endpoint> self;
     };
 
     static void send_raw_cb(std::unique_ptr<threading::Tmsg<SendRecvMsg>> msg)
     {
-      reinterpret_cast<TLSEndpoint*>(msg->data.self.get())
-        ->send_raw_thread(msg->data.data);
+      auto ptr = msg->data.self.lock();
+      reinterpret_cast<TLSEndpoint*>(ptr.get())->send_raw_thread(
+        msg->data.data);
```

Comment:
```suggestion
      auto ptr = msg->data.self.lock();
      if (ptr == nullptr)
      {
        throw std::logic_error("Executing SendRecvMsg on a dead Endpoint");
      }
      reinterpret_cast<TLSEndpoint*>(ptr.get())->send_raw_thread(
        msg->data.data);
```

Reply:
I literally have that locally :)

For some reason though, tls sessions seem to die early, trying to work out why.

Reply:
Also while we're here - we can remove the `reintepret_cast` if we store `weak_ptr<TLSEndpoint>`. `TLSEndpoint` (and `HTTPEndpoint`) should inherit directly from `std::enable_shared_from_this` (rather than via `Endpoint`).

So
`class TLSEndpoint : public Endpoint, public std::enable_shared_from_this<TLSEndpoint>`
`weak_ptr<TLSEndpoint> self;`
`ptr->send_raw_thread(...`